Explore the option of using the "strict" version

It seems the "strict" version of the schema might yield better results with respect to a Regex error in earlier versions of the JSON schema.

The default setting causes `kotlinx.json` to throw a `JsonConvertException` like so:
```
Caused by: io.ktor.serialization.JsonConvertException: Illegal input: Unexpected JSON token at offset 23: Encountered an unknown key 'foo' at path: $.document
Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.
JSON input: {
  "document": {
    "foo": "bar",
    "aggregate_se.....
	at _COROUTINE._BOUNDARY._(CoroutineDebugging.kt:42)
	at io.ktor.client.HttpClient$4.invokeSuspend(HttpClient.kt:177)
	at io.ktor.client.plugins.HttpCallValidator$Companion$install$2.invokeSuspend(HttpCallValidator.kt:142)
	at io.ktor.client.call.HttpClientCall.bodyNullable(HttpClientCall.kt:89)
	at io.github.csaf.sbom.retrieval.CsafLoader$fetchDocument$2.invokeSuspend(CsafLoader.kt:170)
```

Thus, we already have the default-fail behavior as intended.

If additional properties are to be allowed, this can be accomplished by setting `ignoreUnknownKeys = true` in `JsonBuilder`.

--- a/csaf-import/src/test/kotlin/io/github/csaf/sbom/retrieval/CsafLoaderTest.kt
+++ b/csaf-import/src/test/kotlin/io/github/csaf/sbom/retrieval/CsafLoaderTest.kt
@@ -16,12 +16,15 @@
  */
 package io.github.csaf.sbom.retrieval
 
+import io.github.csaf.sbom.schema.generated.Csaf
 import io.github.csaf.sbom.validation.ValidationContext
 import io.github.csaf.sbom.validation.ValidationException
 import io.ktor.http.*
-import kotlin.test.*
 import kotlinx.coroutines.test.runTest
-import org.junit.jupiter.api.assertThrows
+import kotlinx.serialization.json.Json
+import kotlin.io.path.Path
+import kotlin.io.path.readText
+import kotlin.test.*
 
 class CsafLoaderTest {
     private val loader = CsafLoader(mockEngine())
@@ -57,7 +60,7 @@ class CsafLoaderTest {
             "Failed to \"download\" example-01-aggregator.json from resources."
         )
         // Fresh [ValidationContext] should always throw.
-        assertThrows<ValidationException> {
+        assertFailsWith<ValidationException> {
             RetrievedProvider(result.getOrThrow()).validate(ValidationContext())
         }
 
@@ -104,4 +107,5 @@ class CsafLoaderTest {
             }
         assertFalse { result.isSuccess }
     }
+
 }

--- a/csaf-import/src/test/kotlin/io/github/csaf/sbom/retrieval/CsafLoaderTest.kt
+++ b/csaf-import/src/test/kotlin/io/github/csaf/sbom/retrieval/CsafLoaderTest.kt
@@ -16,15 +16,11 @@
  */
 package io.github.csaf.sbom.retrieval
 
-import io.github.csaf.sbom.schema.generated.Csaf
 import io.github.csaf.sbom.validation.ValidationContext
 import io.github.csaf.sbom.validation.ValidationException
 import io.ktor.http.*
-import kotlinx.coroutines.test.runTest
-import kotlinx.serialization.json.Json
-import kotlin.io.path.Path
-import kotlin.io.path.readText
 import kotlin.test.*
+import kotlinx.coroutines.test.runTest
 
 class CsafLoaderTest {
     private val loader = CsafLoader(mockEngine())
@@ -107,5 +103,4 @@ class CsafLoaderTest {
             }
         assertFalse { result.isSuccess }
     }
-
 }

--- a/csaf-schema/src/test/kotlin/io/github/csaf/sbom/schema/CsafTest.kt
+++ b/csaf-schema/src/test/kotlin/io/github/csaf/sbom/schema/CsafTest.kt
@@ -22,9 +22,27 @@ import java.net.URI
 import java.time.Instant
 import java.time.OffsetDateTime
 import java.time.ZoneOffset
+import kotlin.io.path.Path
+import kotlin.io.path.readText
 import kotlin.test.*
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.json.Json
 
 class CsafTest {
+
+    @Test
+    fun testStrict() {
+        val file = Path("src/test/resources/csaf-strict-test.json")
+        val ex =
+            assertFailsWith<SerializationException> { Json.decodeFromString<Csaf>(file.readText()) }
+        val msg = ex.message
+        assertNotNull(msg)
+        assertTrue(
+            msg.contains(
+                "Unexpected JSON token at offset 5: Encountered an unknown key 'this_property_does_not_exist' at path"
+            )
+        )
+    }
 
     @Test
     fun testFailCVSSVector() {

--- a/csaf-schema/src/test/resources/csaf-strict-test.json
+++ b/csaf-schema/src/test/resources/csaf-strict-test.json
@@ -14,9 +14,9 @@
     },
     "lang": "en-US",
     "publisher": {
-      "category": "coordinator",
-      "name": "Bundesamt für Sicherheit in der Informationstechnik",
-      "namespace": "https://www.bsi.bund.de"
+      "category": "other",
+      "name": "csaf-sbom/kotlin-csaf Test files",
+      "namespace": "https://csaf-sbom.github.io/"
     },
     "title": "CVRF-CSAF-Converter: XML External Entities Vulnerability",
     "tracking": {
@@ -28,7 +28,7 @@
           "version": "1.12.1"
         }
       },
-      "id": "BSI-2022-0001",
+      "id": "CSAF-Strict-Test",
       "initial_release_date": "2022-03-17T13:03:42.105Z",
       "revision_history": [
         {

--- a/csaf-schema/src/test/resources/csaf-strict-test.json
+++ b/csaf-schema/src/test/resources/csaf-strict-test.json
@@ -1,0 +1,215 @@
+{
+  "this_property_does_not_exist": {},
+  "document": {
+    "aggregate_severity": {
+      "text": "Moderate"
+    },
+    "category": "csaf_security_advisory",
+    "csaf_version": "2.0",
+    "distribution": {
+      "tlp": {
+        "label": "WHITE",
+        "url": "https://www.first.org/tlp/"
+      }
+    },
+    "lang": "en-US",
+    "publisher": {
+      "category": "coordinator",
+      "name": "Bundesamt für Sicherheit in der Informationstechnik",
+      "namespace": "https://www.bsi.bund.de"
+    },
+    "title": "CVRF-CSAF-Converter: XML External Entities Vulnerability",
+    "tracking": {
+      "current_release_date": "2022-03-17T13:03:42.105Z",
+      "generator": {
+        "date": "2022-03-17T13:09:42.105Z",
+        "engine": {
+          "name": "Secvisogram",
+          "version": "1.12.1"
+        }
+      },
+      "id": "BSI-2022-0001",
+      "initial_release_date": "2022-03-17T13:03:42.105Z",
+      "revision_history": [
+        {
+          "date": "2022-03-17T13:03:42.105Z",
+          "number": "1",
+          "summary": "Initial revision"
+        }
+      ],
+      "status": "final",
+      "version": "1"
+    }
+  },
+  "product_tree": {
+    "branches": [
+      {
+        "branches": [
+          {
+            "branches": [
+              {
+                "category": "product_version",
+                "name": "1.0.0-alpha",
+                "product": {
+                  "name": "CSAF Tools CVRF-CSAF-Converter 1.0.0-alpha",
+                  "product_id": "CSAFPID-0001",
+                  "product_identification_helper": {
+                    "cpe": "cpe:/a:csaf-tools:cvrf-csaf-converter:1.0.0-alpha"
+                  }
+                }
+              },
+              {
+                "category": "product_version",
+                "name": "1.0.0-dev1",
+                "product": {
+                  "name": "CSAF Tools CVRF-CSAF-Converter 1.0.0-dev1",
+                  "product_id": "CSAFPID-0002",
+                  "product_identification_helper": {
+                    "cpe": "cpe:/a:csaf-tools:cvrf-csaf-converter:1.0.0-dev1"
+                  }
+                }
+              },
+              {
+                "category": "product_version",
+                "name": "1.0.0-dev2",
+                "product": {
+                  "name": "CSAF Tools CVRF-CSAF-Converter 1.0.0-dev2",
+                  "product_id": "CSAFPID-0003",
+                  "product_identification_helper": {
+                    "cpe": "cpe:/a:csaf-tools:cvrf-csaf-converter:1.0.0-dev2"
+                  }
+                }
+              },
+              {
+                "category": "product_version",
+                "name": "1.0.0-dev3",
+                "product": {
+                  "name": "CSAF Tools CVRF-CSAF-Converter 1.0.0-dev3",
+                  "product_id": "CSAFPID-0004",
+                  "product_identification_helper": {
+                    "cpe": "cpe:/a:csaf-tools:cvrf-csaf-converter:1.0.0-dev3"
+                  }
+                }
+              },
+              {
+                "category": "product_version",
+                "name": "1.0.0-rc1",
+                "product": {
+                  "name": "CSAF Tools CVRF-CSAF-Converter 1.0.0-rc1",
+                  "product_id": "CSAFPID-0005",
+                  "product_identification_helper": {
+                    "cpe": "cpe:/a:csaf-tools:cvrf-csaf-converter:1.0.0-rc1"
+                  }
+                }
+              },
+              {
+                "category": "product_version",
+                "name": "1.0.0-rc2",
+                "product": {
+                  "name": "CSAF Tools CVRF-CSAF-Converter 1.0.0-rc2",
+                  "product_id": "CSAFPID-0006",
+                  "product_identification_helper": {
+                    "cpe": "cpe:/a:csaf-tools:cvrf-csaf-converter:1.0.0-rc2"
+                  }
+                }
+              }
+            ],
+            "category": "product_name",
+            "name": "CVRF-CSAF-Converter"
+          }
+        ],
+        "category": "vendor",
+        "name": "CSAF Tools"
+      }
+    ]
+  },
+  "vulnerabilities": [
+    {
+      "acknowledgments": [
+        {
+          "names": [
+            "Damian Pfammatter"
+          ],
+          "organization": "Cyber-Defense Campus",
+          "summary": "Finding and reporting the vulnerability"
+        }
+      ],
+      "cve": "CVE-2022-27193",
+      "cwe": {
+        "id": "CWE-611",
+        "name": "Improper Restriction of XML External Entity Reference"
+      },
+      "ids": [
+        {
+          "system_name": "Github Issue",
+          "text": "csaf-tools/CVRF-CSAF-Converter#78"
+        }
+      ],
+      "notes": [
+        {
+          "category": "description",
+          "text": "CSAF Tools CVRF-CSAF-Converter 1.0.0-rc1 resolves XML External Entities (XXE). This leads to the inclusion of arbitrary (local) file content into the generated output document. An attacker can exploit this to disclose information from the system running the converter.",
+          "title": "Vulnerability description"
+        }
+      ],
+      "product_status": {
+        "first_fixed": [
+          "CSAFPID-0006"
+        ],
+        "fixed": [
+          "CSAFPID-0006"
+        ],
+        "known_affected": [
+          "CSAFPID-0001",
+          "CSAFPID-0002",
+          "CSAFPID-0003",
+          "CSAFPID-0004",
+          "CSAFPID-0005"
+        ]
+      },
+      "remediations": [
+        {
+          "category": "vendor_fix",
+          "date": "2022-03-14T13:10:55.000+01:00",
+          "details": "Update to the latest version of the product. At least version 1.0.0-rc2",
+          "product_ids": [
+            "CSAFPID-0001",
+            "CSAFPID-0002",
+            "CSAFPID-0003",
+            "CSAFPID-0004",
+            "CSAFPID-0005"
+          ],
+          "url": "https://github.com/csaf-tools/CVRF-CSAF-Converter/releases/tag/1.0.0-rc2"
+        }
+      ],
+      "scores": [
+        {
+          "cvss_v3": {
+            "attackComplexity": "LOW",
+            "attackVector": "LOCAL",
+            "availabilityImpact": "LOW",
+            "baseScore": 6.1,
+            "baseSeverity": "MEDIUM",
+            "confidentialityImpact": "HIGH",
+            "exploitCodeMaturity": "FUNCTIONAL",
+            "integrityImpact": "NONE",
+            "privilegesRequired": "NONE",
+            "remediationLevel": "OFFICIAL_FIX",
+            "reportConfidence": "CONFIRMED",
+            "scope": "UNCHANGED",
+            "userInteraction": "REQUIRED",
+            "vectorString": "CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:U/C:H/I:N/A:L/E:F/RL:O/RC:C",
+            "version": "3.1"
+          },
+          "products": [
+            "CSAFPID-0001",
+            "CSAFPID-0002",
+            "CSAFPID-0003",
+            "CSAFPID-0004",
+            "CSAFPID-0005"
+          ]
+        }
+      ]
+    }
+  ]
+}

